plugins/alerting/pkg/alerting/api_endpoints.go

```diff
@@ -0,0 +1,250 @@
+package alerting
+
+import (
+	"context"
+	"fmt"
+	"path"
+
+	"github.com/google/uuid"
+	cfg "github.com/prometheus/alertmanager/config"
+	"github.com/rancher/opni/pkg/alerting/shared"
+	alertingv1alpha "github.com/rancher/opni/pkg/apis/alerting/v1alpha"
+	corev1 "github.com/rancher/opni/pkg/apis/core/v1"
+	"github.com/rancher/opni/pkg/validation"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+const endpointPrefix = "/alerting/endpoints"
+
+func configFromBackend(backend RuntimeEndpointBackend, ctx context.Context, p *Plugin) (*ConfigMapData, error) {
+	rawConfig, err := backend.Fetch(ctx, p, "alertmanager.yaml")
+	if err != nil {
+		return nil, err
+	}
+	config, err := NewConfigMapDataFrom(rawConfig)
+	if err != nil {
+		return nil, err
+	}
+	return config, nil
+}
+
+func applyConfigToBackend(backend RuntimeEndpointBackend, ctx context.Context, p *Plugin, config *ConfigMapData) error {
+	newRawConfig, err := config.Marshal()
+	if err != nil {
+		return err
+	}
+	err = backend.Put(ctx, p, "alertmanager.yaml", string(newRawConfig))
+	if err != nil {
+		return err
+	}
+	err = backend.Reload(ctx, p)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func validateSlack(v *alertingv1alpha.SlackEndpoint) error {
+	if v == nil {
+		return validation.Error("Must pass in a non-nil slack endpoint")
```

Comment:
For consistency, use a message like `Missing required field:`

Reply:
Fixed